### DISCUSSION TITLE
Multiple instances of StateNotifier are created

### DISCUSSION THREAD

**Discussion by alextran1502** [rrousselGit/riverpod#1191] [Q&A] [answered]
**Describe the bug**
I have a WebSocket StateNotifier to receive WebSocket's event and it has `Ref ref` as a parameter to communicate with different providers. Below is the StateNotifier class and the provider variable.

```dart
class WebsocketNotifier extends StateNotifier<WebscoketState> {
  WebsocketNotifier(this.ref) : super(WebscoketState(socket: null, isConnected: false)) {
    debugPrint("Init websocket instance");
  }

  final Ref ref;

  connect() {
    var authenticationState = ref.watch(authenticationProvider);

    if (authenticationState.isAuthenticated) {
      var accessToken = Hive.box(userInfoBox).get(accessTokenKey);
      var endpoint = Hive.box(userInfoBox).get(serverEndpointKey);
      try {
        debugPrint("[WEBSOCKET] Attempting to connect to ws");
        // Configure socket transports must be sepecified
        Socket socket = io(
          endpoint,
          OptionBuilder()
              .setTransports(['websocket'])
              .enableReconnection()
              .enableForceNew()
              .enableForceNewConnection()
              .enableAutoConnect()
              .setExtraHeaders({"Authorization": "Bearer $accessToken"})
              .build(),
        );

        socket.onConnect((_) {
          debugPrint("[WEBSOCKET] Established Websocket Connection");
          state = WebscoketState(isConnected: true, socket: socket);
        });

        socket.onDisconnect((_) {
          debugPrint("[WEBSOCKET] Disconnect to Websocket Connection");
          state = WebscoketState(isConnected: false, socket: null);
        });

        socket.on('error', (errorMessage) {
          debugPrint("Webcoket Error - $errorMessage");
          state = WebscoketState(isConnected: false, socket: null);
        });

        socket.on('on_upload_success', (data) {
          var jsonString = jsonDecode(data.toString());
          ImmichAsset newAsset = ImmichAsset.fromMap(jsonString);
          ref.watch(assetProvider.notifier).onNewAssetUploaded(newAsset);
        });
      } catch (e) {
        debugPrint("[WEBSOCKET] Catch Websocket Error - ${e.toString()}");
      }
    }
  }

  disconnect() {
    debugPrint("[WEBSOCKET] Attempting to disconnect");
    var socket = state.socket?.disconnect();
    if (socket != null) {
      if (socket.disconnected) {
        state = WebscoketState(isConnected: false, socket: null);
      }
    }
  }
}

final websocketProvider = StateNotifierProvider<WebsocketNotifier, WebscoketState>((ref) {
  return WebsocketNotifier(ref);
});

```
I have a drawer that handles logging out by calling ` ref.watch(websocketProvider.notifier).disconnect();`. It always creates a new instance of the StateNotifier, therefore cannot log out the instance I want. I believe I made a mistake somewhere and what caused this behavior?

**To Reproduce**

You can clone my repository and run the application - I have a dockerized the application so it is fairly straightforward to replicate 
https://github.com/alextran1502/immich

**Expected behavior**
Using the same instance Websocket to log out.

**Comment by rrousselGit** [comment, marked as answer]
You're using ref.watch in a couple of places within your class. If the value obtained changes, that'll recreate the class.

Maybe you want to use ref.listen instead

**Reply by alextran1502** [reply]
Where should I use `ref.listen`? Here is my `onTap` method from ListTile that handles logging out and cleaning up the application. I want to invoke the method in `StateNotifier`, and the method will change the state of the provider.

![image](https://user-images.githubusercontent.com/27055614/153913880-1091bc90-a24b-430c-b94e-149eff6f40fc.png)

Thank you for your help. I appreciate your time

**Reply by rrousselGit** [reply]
I'm not referring to your widgets here. I'm talking about within the StateNotifier

You are using ref.watch within the StateNotifier. Change those to ref.listen

**Reply by alextran1502** [reply]
AH, I see! The authentication state changes after the logout method is executed, which probably leads to the recreation of the class. Thank you for your insight!
